src/ray/raylet/node_manager.cc

```diff
@@ -790,6 +790,13 @@ void NodeManager::TreatTaskAsFailed(const TaskSpecification &spec) {
 
 void NodeManager::SubmitTask(const Task &task, const Lineage &uncommitted_lineage,
                              bool forwarded) {
+  if (local_queues_.TaskInQueue(task.GetTaskSpecification().TaskId())) {
+    RAY_LOG(WARNING)
+        << "Submitted task " << task.GetTaskSpecification().TaskId()
+        << " is already queued. This is most likely due to spurious reconstruction.";
+    return;
+  }
```

Comment:
Is this a performance concern? It seems like a lot of dictionary lookups for every task.

Reply:
I think it is probably okay since it's a constant number. If this is an issue, we can also add an `unordered_set` to the `SchedulingQueue` that contains all queued tasks. We would probably have to rethink the API for the `SchedulingQueue`, though, to make sure that we update that data structure properly.